lua/gitlab/reviewer/diffview.lua

```diff
@@ -126,106 +126,117 @@ M.get_location = function(range)
+  local b_win = u.get_window_id_by_buffer_id(layout.b.file.bufnr)
+  local current_win = vim.fn.win_getid()
+  local is_current_sha = current_win == b_win
+
+  if a_win == nil or b_win == nil then
+    u.notify("Error retrieving window IDs for current files", vim.log.levels.ERROR)
     return
   end
 
-  local hunks = u.parse_hunk_headers(result.file_name, state.INFO.target_branch)
-  if hunks == nil then
-    u.notify("Could not parse hunks", vim.log.levels.ERROR)
+  local current_file = M.get_current_file()
+  if current_file == nil then
+    u.notify("Error retrieving current file from Diffview", vim.log.levels.ERROR)
     return
   end
 
-  local current_line_info
-  if is_new then
-    current_line_info = u.get_lines_from_hunks(hunks, result.new_line, is_new)
-  else
-    current_line_info = u.get_lines_from_hunks(hunks, result.old_line, is_new)
+  local a_linenr = vim.api.nvim_win_get_cursor(a_win)[1]
+  local b_linenr = vim.api.nvim_win_get_cursor(b_win)[1]
+
+  local data = u.parse_hunk_headers(current_file, state.INFO.target_branch)
+
+  if data.hunks == nil then
+    u.notify("Could not parse hunks", vim.log.levels.ERROR)
+    return
   end
 
-  -- If single line comment is outside of changed lines then we need to specify both new line and old line
-  -- otherwise the API returns error.
-  -- https://docs.gitlab.com/ee/api/discussions.html#create-a-new-thread-in-the-merge-request-diff
-  if not current_line_info.in_hunk then
-    result.old_line = current_line_info.old_line
-    result.new_line = current_line_info.new_line
+  -- Will be different depending on focused window.
+  local modification_type =
+    M.get_modification_type(a_linenr, b_linenr, is_current_sha, data.hunks, data.all_diff_output)
+
+  if modification_type == "bad_file_unmodified" then
+    u.notify("Comments on unmodified lines will be placed in the old file", vim.log.levels.WARN)
   end
 
-  -- If users leave single-line comments in the new buffer that should be in the old buffer, we can
-  -- tell because the line will not have changed. Send the correct payload.
-  if M.lines_are_same(view.cur_layout) and layout.b.file.bufnr == bufnr and range == nil then
-    local a_win = u.get_win_from_buf(layout.a.file.bufnr)
-    local a_cursor = vim.api.nvim_win_get_cursor(a_win)[1]
-    result.old_line = a_cursor
-    result.new_line = a_cursor
-    type = "old"
+  -- Comment on new line: Include only new_line in payload.
+  if modification_type == "added" then
+    reviewer_info.old_line = nil
+    reviewer_info.new_line = b_linenr
+    -- Comment on deleted line: Include only new_line in payload.
+  elseif modification_type == "deleted" then
+    reviewer_info.old_line = a_linenr
+    reviewer_info.new_line = nil
+    -- The line was not found in any hunks, only send the old line number
+  elseif modification_type == "unmodified" or modification_type == "bad_file_unmodified" then
+    reviewer_info.old_line = a_linenr
+    reviewer_info.new_line = b_linenr
```

Comment:
Here we're actually building the payload correctly depending on what the helper `get_modification_type` returns.